Core/include/Acts/Vertexing/NumericalTrackLinearizer.hpp

```diff
@@ -0,0 +1,164 @@
+///
+/// @tparam propagator_t Propagator type
+/// @tparam propagator_options_t Propagator options type
+template <typename propagator_t,
+          typename propagator_options_t = PropagatorOptions<>>
+class NumericalTrackLinearizer {
+ public:
+  using Propagator_t = propagator_t;
+
+  /// State struct
+  struct State {
+    /// @brief State constructor
+    ///
+    /// @param fieldCacheIn Magnetic field cache
+    State(MagneticFieldProvider::Cache fieldCacheIn)
+        : fieldCache(std::move(fieldCacheIn)) {}
+    MagneticFieldProvider::Cache fieldCache;
+  };
+
+  /// @brief Configuration struct
+  struct Config {
+    /// @ Config constructor if magnetic field is present
+    ///
+    /// @param bIn The magnetic field
+    /// @param prop The propagator
+    Config(std::shared_ptr<const MagneticFieldProvider> bIn,
+           std::shared_ptr<const Propagator_t> prop)
+        : bField(std::move(bIn)), propagator(std::move(prop)) {}
+
+    /// @brief Config constructor without B field -> uses NullBField
+    ///
+    /// @param prop Propagator
+    Config(std::shared_ptr<const Propagator_t> prop)
+        : bField{std::make_shared<NullBField>()}, propagator(std::move(prop)) {}
+
+    std::shared_ptr<const MagneticFieldProvider> bField;
+
+    std::shared_ptr<const Propagator_t> propagator;
+
+    // Tolerance determining how close we need to get to a surface to
+    // reach it during propagation
+    ActsScalar targetTolerance = 1e-12;
+
+    // Setting size of the perturbation delta for calculation of numerical
+    // derivatives (i.e., f'(x) ~ (f(x+delta) - f(x)) / delta)
+    ActsScalar delta = 1e-8;
+  };
+
+  /// @brief Constructor
+  ///
+  /// @param config Configuration object
+  /// @param _logger Logger instance
+  NumericalTrackLinearizer(const Config& config,
+                           std::unique_ptr<const Logger> _logger =
+                               getDefaultLogger("NumTrkLinProp", Logging::INFO))
+      : m_cfg(config), m_logger{std::move(_logger)} {}
+
+  /// @brief Function that linearizes BoundTrackParameters at
+  /// the PCA to a given Perigee surface
+  ///
+  /// @param params Parameters to linearize
+  /// @param linPoint Point which defines the Perigee.
+  /// @note Transverse plane of the Perigee corresponding to @p linPoint is
+  /// parallel to the global x-y plane
+  /// @param gctx Geometry context
+  /// @param mctx Magnetic field context
+  /// @param state Linearizer state object
+  ///
+  /// @return Linearized track
+  Result<LinearizedTrack> linearizeTrack(const BoundTrackParameters& params,
+                                         const Vector4& linPoint,
+                                         const Acts::GeometryContext& gctx,
+                                         const Acts::MagneticFieldContext& mctx,
+                                         State& state) const {
+    // Using "state" to suppress unused-parameter warning
+    (void)state;
```

Comment:
```suggestion
                                         State& /*state*/) const {
```
would be an alternative.